cluster_experiments/experiment_analysis.py

```diff
@@ -680,3 +683,181 @@ def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> fl
+        time_col: str = "date",
+        transition_date: str = None,
+    ):
+        super().__init__(cluster_cols)
+
+        self.target_col = target_col
+        self.treatment = treatment
+        self.treatment_col = treatment_col
+        self.cluster_cols = cluster_cols
+        self.hypothesis = hypothesis
+        self.time_col = time_col
+        self.transition_date = transition_date
+
+    def fit_synthetic(self, df: pd.DataFrame) -> list:
+        """Returns the fitted OLS model"""
+
+        # todo should I throw an error if more than one cluster column is passed?
+
+        X = (
+            df.query(f"{self.treatment_col}==0")
+            .pivot(index=self.cluster_cols, columns=self.time_col)[self.target_col]
+            .T
+        )
+
+        if not any(df[self.treatment_col] == 1):
+            raise ValueError("No treatment unit found in the data.")
+
+        y = df.query(f"{self.treatment_col}==1")[f"{self.target_col}"]  # treated
+
+        weights = get_w(X, y)
+
+        return weights
+
+    def synthetic_control(self, df: pd.DataFrame, treatment_unit) -> np.array:
+        weights = self.fit_synthetic(df)
+
+        synthetic = (
+            df.query(
+                f"{''.join(self.cluster_cols)} != '{treatment_unit}'"
+            )  # remove treatment cluster
+            .pivot(index=self.time_col, columns=self.cluster_cols)[self.target_col]
+            .values.dot(weights)
+        )
+
+        return df.query(
+            f"{''.join(self.cluster_cols)} == '{treatment_unit}'"
+        ).assign(  # add synthetic to treatment unit
+            synthetic=synthetic
+        )
+
+    def pvalue_based_on_hypothesis(
+        self, synthetic_donors: pd.DataFrame, treatment_cluster: str
+    ) -> float:
+
+        treatment_index = [
+            i for i, df in enumerate(synthetic_donors) if (df["treatment"] == 1).any()
+        ]
+
+        # first calculate the average effect after intervention for each unit, in a list comprehension
+        avg_effects = [
+            unit.query("treatment_period == 'After'")[self.target_col].mean()
+            for unit in synthetic_donors
+        ]
+        avg_effects = np.abs(avg_effects)
+
+        # then count how many times the average effect is greater than the real treatment unit (which is the first one)
+        # and divide by the number of units
+        p_value = np.where(avg_effects > avg_effects[treatment_index], 1, 0).mean()
+
+        return p_value
+
+    def analysis_pvalue(self, df: pd.DataFrame, verbose: bool = False) -> float:
+        """Returns the p-value of the analysis
+        Arguments:
+            df: dataframe containing the data to analyze
+            verbose (Optional): bool, prints the regression summary if True
+        """
+
+        # todo should I throw an error if more than one cluster column is passed?
+        control_pool = df[self.cluster_cols].iloc[:, 0].unique()
+        # control_pool = df[self.cluster_cols].iloc[:, 0].values
```

Comment:
I will leave this for now because this is a working version with non parallel execution. Once the PR is ready for review (not draft) I will exclude any commented code